roboteam_ai/src/coach/midField/HarassRobotCoach.cpp

```diff
@@ -0,0 +1,247 @@
+        double lengthSquared = (robotPos - enemyRobot->pos).length2();
+        if (lengthSquared < closestDistanceSquared) {
+            closestDistanceSquared = lengthSquared;
+            bestIndex = i;
+        }
+    }
+    return bestIndex;
+}
+
+Vector2 HarassRobotCoach::harassRobot(int myIndex, int id, bool stayInMidField) {
+    // (re)set the robot harassed to the robot with that id
+    targetRobotsToHarass[myIndex] = RobotPtr(nullptr);
+    RobotPtr robotToHarass = world::world->getRobotForId(id, false);
+    BallPtr ball = world::world->getBall();
+    if (! robotToHarass) return {currentRobotPositions[myIndex]};
+    if (! ball) return {robotToHarass->pos.x - 0.3, robotToHarass->pos.y};
+
+    // set target
+    Vector2 target;
+    if (robotToHarass->getDistanceToBall() >= 0.0) {
+        target = ball->pos + (ball->pos - robotToHarass->pos).stretchToLength(0.2);
+    }
+    else {
+        double a = 0.74;
+        target = robotToHarass->pos*a + ball->pos*(1-a);
+        target.x = target.x > bestXPos + 1.0 ? target.x*a + (bestXPos + 1.0)*(1-a) : target.x;
+        target.x = target.x < bestXPos - 1.0 ? target.x*a + (bestXPos - 1.0)*(1-a) : target.x;
+    }
+
+    targetRobotPositions[myIndex] = target;
+    targetRobotsToHarass[myIndex] = robotToHarass;
+    for (int i = 0; i < targetRobotsToHarass.size(); i++) {
+        if (i == myIndex) continue;
+        auto &targetRobot = targetRobotsToHarass[i];
+        if (! targetRobot) continue;
+
+        if (targetRobot->id == robotToHarass->id) {
+            targetRobot = RobotPtr(nullptr);
+        }
+    }
+    return target;
+}
+
+Vector2 HarassRobotCoach::standFree(const RobotPtr &thisRobot, int myIndex, const RobotPtr &ourRobotWithBall) {
+
+    Vector2 currentLocation = thisRobot->pos;
+    Vector2 passLine = ourRobotWithBall->pos - currentLocation;
+    Angle passAngle = passLine.toAngle();
+    Angle smallestAngle = Angle(M_PI);
+    auto enemyRobots = world::world->getThem();
+    for (auto &robot : enemyRobots) {
+        Vector2 blockLine = ourRobotWithBall->pos - robot.pos;
+        Angle blockAngle = blockLine.toAngle();
+        Angle deltaAngle = blockAngle - passAngle;
+        if (abs(deltaAngle.getAngle()) < abs(smallestAngle.getAngle())) {
+            auto distToLine = control::ControlUtils::distanceToLine(robot.pos, ourRobotWithBall->pos, currentLocation);
+            auto distToLineWithEnds = control::ControlUtils::distanceToLineWithEnds(robot.pos, ourRobotWithBall->pos, currentLocation);
+            if (distToLine == distToLineWithEnds) {
+                smallestAngle = deltaAngle;
+            }
+        }
+    }
+    Vector2 target;
+    Vector2 closeRobotPos = Vector2(42.0, 42.0);
+    for (auto &otherRobot : world::world->getAllRobots()) {
+        if (otherRobot.id == thisRobot->id && otherRobot.team == thisRobot->team) continue;
+
+        if ((otherRobot.pos - currentLocation).length2() < 0.5) {
+            closeRobotPos = otherRobot.pos;
```

Comment:
so this just randomly picks a robot that is within sqrt(0.5) distance of the point? Might want to explain what/why you are doing this